File: subprojects/cli/commands/run-calls/src/scripts.js
import fs from "node:fs";
import path from "node:path";
import { spawn } from "node:child_process";
import { strapConfigRoot, strapGlobalRoot, strapProjectRoot, strapRoot, strapSessionRoot, strapToolDirs, strapWorkRoot, workspaceRoot } from "./paths.js";
import { toolResult, truncateText } from "./result.js";

const defaultSchema = { type: "object", properties: {}, additionalProperties: true };

function safeName(fileName) {
  return path.basename(fileName).replace(/\.[^.]+$/, "").replace(/[^a-zA-Z0-9_]/g, "_");
}

function readJson(filePath) {
  return fs.existsSync(filePath) ? JSON.parse(fs.readFileSync(filePath, "utf8")) : undefined;
}

function executable(filePath) {
  try {
    fs.accessSync(filePath, fs.constants.X_OK);
    return fs.statSync(filePath).isFile();
  } catch {
    return false;
  }
}

function discover() {
  const seen = new Set();
  const files = [];
  for (const dir of strapToolDirs()) {
    const resolved = path.resolve(dir);
    if (!fs.existsSync(resolved) || !fs.statSync(resolved).isDirectory()) continue;
    for (const entry of fs.readdirSync(resolved, { withFileTypes: true })) {
      if (!entry.isFile() || entry.name.endsWith(".json")) continue;
      const file = path.join(resolved, entry.name);
      if (executable(file) && !seen.has(file)) {
        seen.add(file);
        files.push(file);
      }
    }
  }
  return files;
}

function specFor(file) {
  const sidecar = readJson(`${file}.json`) || readJson(file.replace(/\.[^.]+$/, ".json")) || {};
  return {
    name: sidecar.name || safeName(file),
    description: sidecar.description || `Run script tool ${path.basename(file)}.`,
    inputSchema: sidecar.inputSchema || sidecar.input_schema || sidecar.parameters || defaultSchema,
    timeoutMs: Number(sidecar.timeoutMs || sidecar.timeout_ms || 120000),
    maxOutputBytes: Number(sidecar.maxOutputBytes || sidecar.max_output_bytes || 60000),
  };
}

function runScript(file, input, spec) {
  return new Promise((resolve, reject) => {
    const child = spawn(file, [], { cwd: workspaceRoot(), env: scriptEnv(spec), stdio: ["pipe", "pipe", "pipe"] });
    const timer = setTimeout(() => child.kill("SIGTERM"), spec.timeoutMs);
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (chunk) => { stdout += chunk; });
    child.stderr.on("data", (chunk) => { stderr += chunk; });
    child.on("error", (error) => { clearTimeout(timer); reject(error); });
    child.on("close", (code, signal) => {
      clearTimeout(timer);
      const limited = truncateText(stderr ? `${stdout}\n[stderr]\n${stderr}` : stdout, spec.maxOutputBytes);
      if (code === 0) resolve(toolResult(limited.text, { script: file, exitCode: code, signal, truncated: limited.truncated }));
      else reject(new Error(`Script tool exited ${code ?? signal}: ${limited.text}`));
    });
    child.stdin.end(`${JSON.stringify(input ?? {})}\n`);
  });
}

function scriptEnv(spec) {
  return { ...process.env, STRAP_ROOT: strapRoot(), STRAP_CONFIG: strapConfigRoot(), STRAP_GLOBAL: strapGlobalRoot(), STRAP_PROJECT: strapProjectRoot(), ...(strapSessionRoot() ? { STRAP_SESSION: strapSessionRoot() } : {}), STRAP_WORK: strapWorkRoot(), STRAP_TOOL_NAME: spec.name, STRAP_WORKSPACE: workspaceRoot() };
}

export function scriptTools() {
  const names = new Set();
  return discover().flatMap((file) => {
    const spec = specFor(file);
    if (names.has(spec.name)) return [];
    names.add(spec.name);
    return [{ name: spec.name, execute: (input) => runScript(file, input, spec) }];
  });
}

File: subprojects/cli/commands/run-calls/src/run_calls/scripts.clj
(ns run-calls.scripts
  (:require [babashka.fs :as fs]
            [babashka.process :as p]
            [cheshire.core :as json]
            [clojure.string :as str]
            [run-calls.common :as c]))

(def default-schema {:type "object" :properties {} :additionalProperties true})

(defn safe-name [file]
  (-> (fs/file-name file) (str/replace #"\.[^.]+$" "") (str/replace #"[^a-zA-Z0-9_]" "_")))

(defn read-json-if-exists [file]
  (when (fs/exists? file) (c/parse-json (slurp file))))

(defn executable? [file]
  (zero? (:exit (p/shell {:continue true :out :string :err :string} "test" "-x" (str file)))))

(defn discover []
  (distinct (for [dir (c/tool-dirs)
                  :when (fs/directory? dir)
                  file (fs/list-dir dir)
                  :when (and (fs/regular-file? file) (not (str/ends-with? (str file) ".json")) (executable? file))]
              (str file))))

(defn spec-for [file]
  (let [parsed (fs/split-ext file)
        sibling (str (first parsed) ".json")
        sidecar (or (read-json-if-exists (str file ".json")) (read-json-if-exists sibling) {})]
    {:name (or (:name sidecar) (safe-name file))
     :timeoutMs (long (or (:timeoutMs sidecar) (:timeout_ms sidecar) 120000))
     :maxOutputBytes (long (or (:maxOutputBytes sidecar) (:max_output_bytes sidecar) 60000))}))

(defn script-env [spec]
  (merge (into {} (System/getenv))
         {"STRAP_ROOT" (c/root) "STRAP_CONFIG" (c/config-root) "STRAP_GLOBAL" (c/global-root)
          "STRAP_PROJECT" (c/project-root) "STRAP_WORK" (c/work-root)
          "STRAP_TOOL_NAME" (:name spec) "STRAP_WORKSPACE" (c/workspace)}
         (when-let [s (c/session-root)] {"STRAP_SESSION" s})))

(defn run-script [file spec input]
  (let [r (p/shell {:out :string :err :string :in (str (json/generate-string (or input {})) "\n")
                    :continue true :dir (c/workspace) :env (script-env spec)} file)
        output (str (:out r) (when-not (str/blank? (:err r)) (str "\n[stderr]\n" (:err r))))
        limited (c/truncate output (:maxOutputBytes spec))]
    (if (zero? (:exit r))
      (c/tool-result (:text limited) {:script file :exitCode (:exit r) :signal nil :truncated (:truncated limited)})
      (throw (ex-info (str "Script tool exited " (:exit r) ": " (:text limited)) {})))))

(defn tools []
  (loop [files (discover) seen #{} out {}]
    (if-let [file (first files)]
      (let [spec (spec-for file)]
        (if (seen (:name spec))
          (recur (rest files) seen out)
          (recur (rest files) (conj seen (:name spec)) (assoc out (:name spec) (fn [args _] (run-script file spec args))))))
      out)))
